### ISSUE TITLE
Is it possible to modify the access to a private variable/block?

### ISSUE THREAD

**Issue by seteahconnoc** [jmoenig/Snap--Build-Your-Own-Blocks#1996]
Hello. Since there is development regarding the OF block (http://snap.berkeley.edu/snapsource/dev/snap.html), I wonder, if the access to a private variable/block can be modified.

For example: Here are the variables and blocks of "Sprite(2)".
![bildschirmfoto von 2018-01-21 07-53-05](https://user-images.githubusercontent.com/23081344/35191695-b4c5b58e-fe81-11e7-9fa9-f6ddc3c46fa3.png)
Is there a way, that "Sprite" cannot see the private variable "var" of "Sprite(2)"? (The context is a CS course on encapsulation.)

**Comment by jguille2** [comment]
Oh,
I think public/private variables and methods could be interesting.

But for me, we must close our design discussions about local-global ones before (and about OF vs TELL/ASK operation). And this would be a big global change to think later.

**Comment by brianharvey** [comment]
We have, up to now at least, made it a policy not to protect users from themselves (or teachers from their students ☺).  So, for example, the input type settings in the block editor are advisory; we don't prevent you putting a list in a number slot, etc.  Teachers can set up project files with some primitives hidden, but students can always use "show primitives" to undo that.  So I think we'd resist hiding variables or methods.  Sprite-local variables are safe against another object dereferencing them _by accident;_ you have to use the OF block.  But that's as far as we go.

**Comment by jguille2** [comment]
Hi Brian,
Many times, in courses for teachers, I hear the request of protect some aspect of the environment: custom blocks, hidden blocks... Protect? Really they want to forbid their edition. And here, I totally agree with you. I love open source projects! I love our open Snap!

But we can think at some point in public/private variables-procedures, because this is different. Three comments about this:

- We don't forbid anything to our users. They are the programmers. And then, they can change the conditions of their programs. They could make public or private everything and they could change it later.

- In this sense, I think public/private attribute is part of the abstraction. It is not to protect us from our wills... but yes to be able to forget things (like getters and setters, that are checking conditions that I made before... but now I'm not paying attention to them).

- And a last comment, about the possibility to interact between projects. We are doing this in Snap4Arduino using our integrated webserver (Snap projects are listening and answering to http requests). And there are other projects like NetsBlox with the idea of collaborative programming. In these cases, it could be interesting that Snap users (with total control in their own project) could share a piece (like a service ) to interact with other projects.

As I said... I need more time to be able to open this scenario... but I think it is not a disposable idea.

Joan

**Comment by brianharvey** [comment]
Right, this isn't as bad as that.  But I see it as the first step in a wrong direction.  Next would be enforcing input types.  Then declaring variables as CONST.  And before you know it, Snap<i>!</i> becomes Java disguised as Scratch, instead of Scheme disguised as Scratch.

Anyway, I didn't close the issue; I'm just voting no.

**Comment by jmoenig** [comment]
What @brianharvey says.

I remember when I myself was arguing against the OF, when we first introduced it to Scratch ten years ago (it wasn't in the first couple of versions of Scratch), and how I myself was arguing "encapsulation". John's answer back then was that the OF-Block only lets you *get* an attribute, not *set* it. That argument never convinced me, but then, the more I've pondered encapsulation and modularity myself, the more I am convinced that it's something that should not be taught to kids at all, instead, kids will discover at some point how tightly coupled entities can get in the way of scaling a project up. The public / private distinction in my own experience rarely means anything at all, it's just a kludge to discipline oneself. If you *really* care about modularity you don't let two objects share any mutable state. That's how the internet works (as Alan always points out), and that's how you scale truly big systems. But not even so called professional languages go so far as to prohibit two objects sharing the same array.

In Snap's OOP model we want to offer a novel design to let kids play with (slot-wise dynamic) inheritance and polymorphism. The main idea is to *empower* (i.e. to offer *more*) not to *limit*. As @brianharvey once put it Snap wants to delight kids by enabling them to do more rather than teach them the "though-shalt-nots" of us old dudes :-)

**Comment by jguille2** [comment]
Thanks @brianharvey and @jmoenig  for your comments.

I agree with you and I like your ideas and goals to _empower_ and not _limit_.

And about _modularity_ and the possibility of connecting projects, I understand we can put the _limits_ in the connection, not in Snap!
  I write down in my TODO list of Snap4Arduino an implementation to select variables and messages enabled to the webserver. Now, all variables and messages are enabled, and you can not decide what you want to share (like an Internet service).

For me, this can be closed.

**Comment by seteahconnoc** [comment]
As I just said on #1888: It 's quite interesting to see this work in progress. Please, keep in mind, that I posted this issue just out of curiosity. It 's just a wondering in the context of a CS course on encapsulation - neither a suggestion nor a proposal. Obviously, I like "Snap!" and the ideas behind it. It 's good, to remember that it 's about empowering people.

I want to give an example, that explains the origin of my wondering: It 's about abstract data types.
![bildschirmfoto von 2018-01-22 15-23-23](https://user-images.githubusercontent.com/23081344/35225331-4a552038-ff88-11e7-87e9-d1038cab52ef.png)
In this case, a reader directly sees, that the programmer of the queue used the variable "content" for the implementation. Now, it 's the reader's choice to use the variable "content" directly or to use the queue's blocks. It 's just something, I 'll have to keep in mind in the next CS didactics discussion.

**Comment by jmoenig** [comment]
Well, just because you have a variable doesn't mean that variable defines the behavior! It could be just one tiny - even largely irrelevant - part of an object's state. If you provide accessor methods those can offer a real interface, such as checking for expected data types or value ranges.

Just because students *see* something doesn't mean you can't offer a variety of "get ..." or "set ..." or "change ..." methods that affect a single instance variable, or - more often - a set of instance variables.

I've sometimes seen "Encapsulation" being referred to "data hiding", especially in German materials. Bear in mind, however, that when you look at the *code* there should never be any *hiding* involved. With Snap! we're creating a programming language that intentionally blurs the line between "code" and "program", i.e. we want the *code* to be as much the *live program* as possible. Therefore I'm convinced that it's permissible to let the programmer always inspect everything about an object, while offering them various tools of abstraction, such as - now - letting them access local state through getter- and setter-methods.

You know, even Smalltalk lets you do that at the lowest level, so it can't be wrong :-)

**Comment by seteahconnoc** [comment]
Again, thank you for your helpful explanations. Especially, thanks for your thoughts on German material on encapsulation. Actually, I was quite surprised, that my wondering lead to this enlightening discussion on the programmer 's freedom. From my point, the user 's full access to the code of all sprites was always out of question.

**Comment by brianharvey** [comment]
Anyway, for people who really believe in protecting the user from themself, it's not enough to hide the variables; you have to hide methods, too.
